contracts/src/v0.8/functions/tests/1_0_0/FunctionsRouter.t.sol

```diff
@@ -3,116 +3,1346 @@ pragma solidity ^0.8.19;
+    bytes memory err = new bytes(0);
+    uint96 juelsPerGas = 0;
+    uint96 costWithoutCallback = 0;
+    address transmitter = NOP_TRANSMITTER_ADDRESS_1;
+    FunctionsResponse.Commitment memory commitment = s_requestCommitment;
+    // Modify request commitment to have a invalid requestId
+    bytes32 invalidRequestId = bytes32("this does not exist");
+    commitment.requestId = invalidRequestId;
+
+    // topic0 (function signature, always checked), NOT topic1 (false), NOT topic2 (false), NOT topic3 (false), and data (true).
+    bool checkTopic1 = false;
+    bool checkTopic2 = false;
+    bool checkTopic3 = false;
+    bool checkData = true;
+    vm.expectEmit(checkTopic1, checkTopic2, checkTopic3, checkData);
+    emit RequestNotProcessed({
+      requestId: s_requestCommitment.requestId,
+      coordinator: address(s_functionsCoordinator),
+      transmitter: NOP_TRANSMITTER_ADDRESS_1,
+      resultCode: FunctionsResponse.FulfillResult.INVALID_REQUEST_ID
+    });
+
+    (FunctionsResponse.FulfillResult resultCode, uint96 callbackGasCostJuels) = s_functionsRouter.fulfill(
+      response,
+      err,
+      juelsPerGas,
+      costWithoutCallback,
+      transmitter,
+      commitment
+    );
+
+    assertEq(uint(resultCode), uint(FunctionsResponse.FulfillResult.INVALID_REQUEST_ID));
+    assertEq(callbackGasCostJuels, 0);
+  }
+
+  function test_Fulfill_RequestNotProcessedInvalidCommitment() public {
+    // Send as committed Coordinator
+    vm.stopPrank();
+    vm.startPrank(address(s_functionsCoordinator));
+
+    bytes memory response = bytes("hello world!");
+    bytes memory err = new bytes(0);
+    uint96 juelsPerGas = 0;
+    uint96 costWithoutCallback = 0;
+    address transmitter = NOP_TRANSMITTER_ADDRESS_1;
+    FunctionsResponse.Commitment memory commitment = s_requestCommitment;
+    // Modify request commitment to have charge more than quoted
+    commitment.estimatedTotalCostJuels = 10 * JUELS_PER_LINK; // 10 LINK
+
+    // topic0 (function signature, always checked), NOT topic1 (false), NOT topic2 (false), NOT topic3 (false), and data (true).
+    bool checkTopic1 = false;
+    bool checkTopic2 = false;
+    bool checkTopic3 = false;
+    bool checkData = true;
+    vm.expectEmit(checkTopic1, checkTopic2, checkTopic3, checkData);
+    emit RequestNotProcessed({
+      requestId: s_requestCommitment.requestId,
+      coordinator: address(s_functionsCoordinator),
+      transmitter: NOP_TRANSMITTER_ADDRESS_1,
+      resultCode: FunctionsResponse.FulfillResult.INVALID_COMMITMENT
+    });
+
+    (FunctionsResponse.FulfillResult resultCode, uint96 callbackGasCostJuels) = s_functionsRouter.fulfill(
+      response,
+      err,
+      juelsPerGas,
+      costWithoutCallback,
+      transmitter,
+      commitment
+    );
+
+    assertEq(uint(resultCode), uint(FunctionsResponse.FulfillResult.INVALID_COMMITMENT));
+    assertEq(callbackGasCostJuels, 0);
+  }
+
+  function test_Fulfill_RequestNotProcessedInsufficientGas() public {
```

Comment:
The commented out test cases are pretty important.  What will is the blocker for getting these working?

Reply:
I'll get everything but the `test_Fulfill_RequestNotProcessedSubscriptionBalanceInvariant` one up, because that one requires some helper contracts to be written